### ISSUE TITLE
Making the order lineitem reference/link to the Voucher Code rather than the Voucher type.

### ISSUE THREAD

**Issue by peteeveleigh** [verbb/gift-voucher#77]
Currently, when purchasing a gift voucher, the line item refers to the Voucher Type and links through to that.

My client wants the line item to show the specific Voucher code that was generated and for the link in the line item to go to that code rather than the voucher type. To be honest, this seems a whole lot more useful to me too.

I'm in the process of writing some custom code to do this but wondered if I was reinventing the wheel or if something was missed when the purchase process was built. Is there, for instance, a way to add the generated code to the line item from the add to cart form?

**Comment by engram-design** [comment]
When you mention "the line item refers to the Voucher Type" do you mean how it's shown in the control panel, or when you loop through line items in an order confirmation template/email/pdf?

A Voucher is the purchasable "thing" you add to your cart (which has a Voucher Type), but a Voucher Code is only created when checkout is completed. A Code records this link by storing `orderId`, `voucherId` and `lineItemId`.

If you're talking about when viewing the order's line items in the control panel, that looks something like:

![image](https://user-images.githubusercontent.com/1221575/141202367-64fc594b-2675-4ce6-b185-c33ca05894f6.png)

Where the link takes you to the Voucher (not the Voucher Type). That's because that's the purchasable thing - the product if you will. While I could update this to point to the new Code generated, there's a complication around quantity. You can purchase 4 Vouchers, being a single line item, and it'll general 4 Codes. It'd be impossible to link to all 4 codes from one line item in the control panel.

If this is a custom template you're building as an order summary/etc, there's nothing stopping you from looping through line items, checking if the line item is a Voucher, and outputting each code generated for that Voucher purchase.

> Is there, for instance, a way to add the generated code to the line item from the add to cart form?

Not really, as when adding to the cart, there is no Code. It's only generated when completing checkout.

If you can let me know the context of what you're doing, that might help?

**Comment by peteeveleigh** [comment]
Hi Josh

You got it. It’s the Line Items shown in the control panel order details - as per your screenshot.
Good point about the quantity though. I can see how that would be a problem,

The context is that my client is seeing the order for a Gift Voucher in the Order detail but they can’t tell which Code it applies to. I think they have some kind of in-house thing going on where they print out and post a voucher so they need to know the Code to use to fulfil the order.

I know they could go the other way, and do it via the Vouchers section of the plugin but that’s an extra step and takes them out of their workflow.

So what we need is to show the CodeKey somewhere in that line item on the order details page. I don’t think it even needs to link (though that would be useful).

I’ve been trying to modify the line item title or add a private note to the line item from the EVENT_AFTER_SAVE gift voucher event. But even though I can see the lineitem data change (by dumping out the line item after a change it), the change doesn’t stick. 

Open to ideas!

Pete

**Comment by engram-design** [comment]
Yeah, that makes sense. We don't have control over the HTML shown in the order edit screen unfortunately, otherwise I'd totally fix that up to output all the codes for that voucher purchase. Commerce is pretty limited by what you can do for a line item's representation due to it being in Vue, and not extensible.

What I might suggest is using the `cp.commerce.order.edit.main-pane` template hook to produce your own table of line items, exactly as your client needs. It'll just be shown below the main table of line items/order summary. We've done this before for other clients with similar needs. It's not amazing to have 2 summaries of line items, but it does help a little.

As for the Private Note, I suppose we could add a comma-delimited set of values for the voucher code. Not exactly user-friendly (and you could edit the line item to remove it), or we could add them to line item options.

![image](https://user-images.githubusercontent.com/1221575/141205872-92a5f069-665b-451f-a6c3-1b0eccc4fbcd.png)

As for using the `EVENT_AFTER_SAVE` Gift Voucher event, that might not work, as you'd probably need to listen to the line item created/populated event to add these notes.

**Comment by peteeveleigh** [comment]
@engram-design After talking to the client they are OK to just have the order referenced in the element index of the Voucher Codes. That's already there so, all good!

However, when searching that list by order reference no results are returned. Should it be possible to search by order reference?

**Comment by engram-design** [comment]
I've just added that column as searchable for the next release. To get this early, change your `verbb/gift-voucher` requirement in `composer.json` to:

```
"require": {
  "verbb/gift-voucher": "dev-craft-3 as 2.6.3",
  "...": "..."
}
```

Then run `composer update`.

Note that you'll also need to run `./craft resave/gift-voucher-codes --update-search-index=1` in order to update the search index with the newly added search attribute, otherwise only new codes will appear in your searches.

**Comment by peteeveleigh** [comment]
That’s awesome. Thanks Josh.

**Comment by peteeveleigh** [comment]
Installed and confirmed working. Thank you.

Now... while you're at it ;-)
Would it be a big job to add an "All Codes" view to the element index? At the moment it's only possible to list codes by their type? Have to say this is only inconvenient due to an early design decision on our part but if it could be added that would be great.

**Comment by engram-design** [comment]
Yep, actually should've been in there right now, fixed!

**Comment by peteeveleigh** [comment]
Hero!

**Comment by engram-design** [comment]
Fixed in [2.7.0](https://github.com/verbb/gift-voucher/releases/tag/2.7.0)
